Provide TrackedStorage primitive in addition (#83)
jelhan/ember-local-storage-decorator: README.md

  }

  updateUser = (name) => {
    this.storage.setItem('user', { name });
  }
}
```

### TrackedStorage API

| Method | Description |
|--------|-------------|
| `getItem(key)` | Retrieve a value |
| `setItem(key, value)` | Store a value (JSON serialized automatically) |
| `removeItem(key)` | Remove a value |
| `clear()` | Clear all items with the same prefix |
| `key(index)` | Get key at index |
| `length` | Number of stored items |
| `clearCache()` | Clear internal cache (useful for testing) |

Values are automatically JSON serialized/deserialized and frozen to prevent mutation. TrackedStorage uses a prefix system (`__tracked_storage__` by default) to isolate its keys from other code using the same storage.

### 2. Pre-instantiated Storage Instances

For convenience, pre-instantiated TrackedStorage instances are provided:

```js
import { trackedLocalStorage, trackedSessionStorage } from 'ember-local-storage-decorator';
import Component from '@glimmer/component';

export default class MyComponent extends Component {
  get currentUser() {
    return trackedLocalStorage.getItem('user');
  }

  saveUser = (user) => {
    trackedLocalStorage.setItem('user', user);

>   }

Reviewer: Let's align the examples to ease reading.

```suggestion
  get user() {
    return trackedLocalStorage.getItem('user');
  }

  updateUser = (user) => {
    trackedLocalStorage.setItem('user', user);
  }
```

}
```

These instances use the default prefix and are ready to use immediately.

### 3. Property Decorators

For a more traditional approach, use the `@localStorage` and `@sessionStorage` decorators to bind class properties directly to storage:

```js
import { localStorage, sessionStorage } from 'ember-local-storage-decorator';
import Component from '@glimmer/component';

export default class MyComponent extends Component {
  @localStorage user;
  @sessionStorage tempData;
}
```

The decorators attach a getter to read the value from storage and a setter to write changes to storage.

> ### Basic Usage

Reviewer: I think this chapter only applies to the decorators. Therefore it needs to be intended one level more.

```suggestion
#### Basic Usage
```

```js
const Klass = class {

>   @localStorage foo;

Reviewer: Let's align with the example above to improve readability.

```suggestion
  @localStorage user;
```

}
const klass = new Klass();

> klass.foo = 'baz';

Reviewer: ```suggestion
klass.user = 'Jane';
```

> klass.foo; // 'baz'

Reviewer: ```suggestion
klass.user; // 'Jane'
```

```

> ### Custom Storage Key

Reviewer: ```suggestion
#### Custom Storage Key
```

You may specify a different key to be used in storage:

```js
const Klass = class {

>   @localStorage('bar') foo;

Reviewer: Aligning the example for readability.

```suggestion
  @localStorage('user') currentUser;
```

};
const klass = new Klass();

> klass.foo = 'baz'; // stored under key 'bar'

Reviewer: Aligning the examples. Also adding the reading from localStorage in another key back for clarity.

```suggestion
klass.currentUser = 'Jane'; // stored under key 'user'
window.localStorage.getItem('user'); // '"Jane"'
```

```

### `@sessionStorage` Decorator

The `@sessionStorage` decorator works identically to `@localStorage` but uses `window.sessionStorage` instead:

```js
const Klass = class {
  @sessionStorage tempData;
};

> ```

Reviewer: The readme does not contain a dedicated chapter for `@localStorage` decorator anymore. The basic usage and custom storage key chapters apply to both decorators. Even though it only uses `@localStorage` to illustrate. Did you maybe removed `@sessionStorage` from those examples by accident?

> ### Default Values

Reviewer: Indentation as above.

```suggestion
#### Default Values
```

You can provide a default value that will be used if no value exists in storage:

```js
const Klass = class {
  @localStorage foo = 'defaultValue';
  @sessionStorage bar = { count: 0 };
};
```

## ⭐ Common Features

All three approaches share these characteristics:

### JSON Serialization
Values are stored as JSON strings. Only values that can be serialized to JSON are supported.

### Deep Freezing

> Objects and arrays are deep frozen to prevent accidental mutation:

Reviewer: ```suggestion

Objects and arrays are deep frozen to prevent accidental mutation:
```

```js
trackedLocalStorage.setItem('data', { items: ['a', 'b'] });
const data = trackedLocalStorage.getItem('data');

Object.isFrozen(data); // true
Object.isFrozen(data.items); // true
```

### Cross-Instance Reactivity

> Changes are automatically observed across different class instances and respond to StorageEvents from other tabs:

Reviewer: Let's give the reader a hint what `StorageEvent` is.

Also missing an empty line after the heading.

```suggestion

Changes are automatically observed across different class instances and respond to [`StorageEvent`s](https://developer.mozilla.org/en-US/docs/Web/API/StorageEvent) from other tabs:
```

```js
const instanceA = new TrackedStorage(window.localStorage);
const instanceB = new TrackedStorage(window.localStorage);

instanceA.setItem('foo', 'bar');
instanceB.getItem('foo'); // 'bar'

// Responds to changes from other browser tabs
window.dispatchEvent(
  new StorageEvent('storage', { 
    key: '__tracked_storage__:foo', 
    newValue: '"baz"' 

>   })

Reviewer: The new implementation which landed in #82 requires the `storageArea` property of the `StorageEvent` to be set, doesn't it?

https://github.com/jelhan/ember-local-storage-decorator/blob/698b18a273af8dd25014645a6900929867e5d723/src/storage/manager.ts#L67-L71

Object.isFrozen(newValue); // false
Object.isFrozen(data); // true
Object.isFrozen(data.items); // true
```

It observes changes caused by other classes or by other instances:
### Cross-Instance Reactivity
Changes are automatically observed across different class instances and respond to StorageEvents from other tabs:

```js
const KlassA = class {
  @localStorage foo;
};
const KlassB = class {
  @localStorage foo;
}
const klassA = new KlassA();
const klassB = new KlassB();
const instanceA = new TrackedStorage(window.localStorage);
const instanceB = new TrackedStorage(window.localStorage);

klassA.foo = 'bar';
klassB.foo; // 'bar'
instanceA.setItem('foo', 'bar');
instanceB.getItem('foo'); // 'bar'

// Responds to changes from other browser tabs
window.dispatchEvent(
  new StorageEvent('storage', { key: 'foo', newValue: 'baz', oldValue: 'bar' })
  new StorageEvent('storage', { 
    key: '__tracked_storage__:foo', 
    newValue: '"baz"' 
  })
);
klassA.foo; // 'baz'
klassB.foo; // 'baz'
instanceA.getItem('foo'); // 'baz'
```

Due to limitations of the Web Storage API, direct changes to the storage 
bypassing the decorator can not be observed. Therefore you _should not_

> manipulate `window.localStorage` or `window.sessionStorage` directly.

Reviewer: This is still relevant, isn't it?

> TrackedStorage uses a prefix system (default: `__tracked_storage__`) to namespace its keys and avoid conflicts with other code using the same storage. The decorators use TrackedStorage internally, so they also benefit from this isolation.

Reviewer: ```suggestion

TrackedStorage uses a prefix system (default: `__tracked_storage__`) to namespace its keys and avoid conflicts with other code using the same storage. The decorators use TrackedStorage internally, so they also benefit from this isolation.
```
